SpinButton migration

misterpekert opened the pull request:
Work in progress, list of components can be found in the issue
Closes #12

edwh commented on components/ChatFooter.vue:
```diff
@@ -133,16 +133,21 @@
-          />
-          <v-icon v-else icon="angle-double-right" title="Send" />
-        </b-button>
+        <spin-button
```
You've got `SpinButton` in some places and `spin-button` in others.  I prefer `SpinButton` - please make consistentn.

edwh commented on components/SpinButton.vue:
```diff
@@ -91,6 +100,33 @@ const props = defineProps({
     required: false,
     default: null,
   },
+  showSpinner: {
```
I'm not convinced that we need `showSpinner`.  Generally:

- The event handler for `handle` can be async.
- The button should only spin while that handler is executing.  
- We should deal with that within this component.
- That means the parent doesn't need a separate local which it sets and resets inside the handler and then passes into this component to make the spinner show.
- This makes the parent code cleaner.  

It's possible that there may be some places where that approach doesn't work, but I think there are some in this PR where it does - `UserSearch` for example.  

So please review the changes and try to remove the local variables in the parents, and see what's left.

misterpekert replied:
You can have async event handlers, sure. What you can't is to have async emits. Emit is a coid function and you can't await the handler which listens for that emit. There are different patterns around that, which we can discuss in a separate issue I've created for refactoring.

edwh replied:
Ah, of course, you're right.

Ok, ignore that feedback - I'll carry on with reviewing.

edwh commented on components/ChatFooter.vue:
```diff
@@ -275,6 +283,7 @@ import { fetchOurOffers } from '../composables/useThrottle'
 import { useAuthStore } from '../stores/auth'
 import { useAddressStore } from '../stores/address'
 import { untwem } from '~/composables/useTwem'
+import SpinButton from "./SpinButton";
```
This line gives a lint error for me.  My IDE (PhpStorm) is set to show lint errors and autofix.  Can you set up to do the same?

There are other examples of this, and it would be nice if your PRs didn't introduce lint errors.

edwh commented on components/DraggableMap.vue:
```diff
@@ -116,6 +118,7 @@ export default {
           this.locating = true
           navigator.geolocation.getCurrentPosition((position) => {
             // Show close to where we think they are.
+            this.locating = false;
```
What's the reason for this change?  It looks like if the geolocate fails we'll end up with `locating` set true forever.

edwh replied:
@misterpekert I think this problem still exists in your latest code.  Please re-check, and also in `Postcode`.

edwh replied:
@misterpekert This is still an issue in the latest code.

misterpekert replied:
I'm notsure what you mean, `navigator.geolocation.getCurrentPosition` has two arguments: `successCallback` and `errorCallback`. I've added `this.locating = false` to both of them. But now, when we don't need this flag, it's enough to call the `callback` method. I've updated the code. I'm not sure what the logic was before, but `navigator.geolocation.getCurrentPosition` never throws an error, so I've added the `this.locationFailed = true` part to the `errorCallback`. Let me know if we should remove try/catch from here.

misterpekert commented:
@edwh I've addressed all the issues you have noticed. As for the linter, I was doing this task from another laptop which for some reason had linter disabled. All linter errors have been addressed and fixed. I think it's ready to be reviewed one more time.

edwh commented:
Thanks.  I'm away this week so I won't get to reviewing that till next 
week.


------ Original Message ------
From "Ilya Pekert" ***@***.***>
To "Freegle/iznik-nuxt3" ***@***.***>
Cc "Edward Hibbert" ***@***.***>; "Mention" 
***@***.***>
Date 07/11/2023 22:08:33
Subject Re: [Freegle/iznik-nuxt3] [WIP]: SpinButton migration (PR #47)

>
>@edwh <https://github.com/edwh> I've addressed all the issues you have 
>noticed. As for the linter, I was doing this task from another laptop 
>which for some reason had linter disabled. All linter errors have been 
>addressed and fixed. I think it's ready to be reviewed one more time.
>
>—
>Reply to this email directly, view it on GitHub 
><https://github.com/Freegle/iznik-nuxt3/pull/47#issuecomment-1800266205>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AAAN26G4AXAYVRJRYPHFD43YDKWODAVCNFSM6AAAAAA6V55DIKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBQGI3DMMRQGU>.
>You are receiving this because you were mentioned.Message ID: 
>***@***.***>
>

edwh commented on components/SettingsPhone.vue:
```diff
@@ -114,26 +106,19 @@ export default {
+    phone() {
       if (this.autoSave) {
-        this.savePhone()
+        this.$refs.spinButton.handle()
```
I'd prefer it if we restructured `savePhone` so that we could call a method without calling into the spin button.  Partly because it's my preference, but also because I see a lot of issues with use of refs which haven't yet been fully loaded, and it would avoid the risk of that.

misterpekert replied:
Oh, yeah, my bad. I've added this for some testing and forgot to remove. SpinButton would never be rendered if autoSave is turned on, so this watch should be left intact. Rolling back that part of the change.

edwh commented on stores/noticeboard.js:
```diff
@@ -85,7 +85,7 @@ export const useNoticeboardStore = defineStore({
     },
     async comments(id, comments) {
-      await this.$api.noticeboard.action({
+      await api(this.config).noticeboard.action({
```
Good spot!

misterpekert replied:
Yeah, took me a while to figure out why it wasn't working as expected :D

edwh reviewed (changes requested):
Getting there, but still some more to do on this one, I'm afraid.

misterpekert commented on components/AutoComplete.vue:
```diff
@@ -20,15 +20,16 @@
+            size="md"
+            button-class="transbord p-0 pr-2 h-100"
+            label=""
+            :spinclass="spinIconClassName"
```
Yeah, good catch. I've wanted to remove SpinButton from here. This component seems to be using only spinner and there is no "action" we can wait for finish.

misterpekert commented on components/SpinButton.vue:
```diff
@@ -1,124 +1,116 @@
     default: 5000,
   },
-  spinclass: {
+  spinColor: {
```
Added the Object with config, should be good. I'd left the prop though just in case it'd be needed in some scenario to override configured color of spinner.

misterpekert commented on components/SpinButton.vue:
```diff
@@ -1,124 +1,116 @@
+    type: String,
+    default: 'fa-fw',
   },
-  confirm: {
```
Well, for me it feels that having the confirmation modal as part of a spinning button a bit "overkill". Any part of the application might need confirmation flow, so from my point of view it should be something standalone. We can think of `<SpinButtonWithConfirm>` component which would have all the perks of SpinButton + confirmation popup, but it felt a bit too much considering the only place of use was `IsoChrone` component.

edwh replied:
As discussed this is going back in.

misterpekert commented on components/IsoChrone.vue:
```diff
@@ -145,20 +145,29 @@
       <hr v-if="!last" class="text-muted mb-1 mt-1" />
     </template>
+    <client-only>
+      <Teleport to="body">
```
Hm, it's just feels a bit more clean and eliminates potential issues with modals overlapping each other if you teleport all of them on the body level. I think if I remove that everything will keep working as with <Teleport>

edwh replied:
I don't understand.  I think Bootstrap already adds the modals to the body level.  There is also the  `no-stacking` prop which controls whether modals can show on top of each other.

So what problem are you trying to solve here?

misterpekert replied:
Well, I didn't know Bootstrap does that by default. I've removed teleport part from SpinButton

misterpekert commented on components/SpinButton.vue:
```diff
@@ -1,124 +1,116 @@
+  props.spinColor ||
+  (props.variant === 'primary' ? ref('text-white') : ref('text-black'))
+
+const finnishSpinner = () => {
```
Fixed that. 🧖‍♂️

edwh commented:
I've found it very useful to have that function to easily add or remove the confirmation stage, and as I say, it's used in the moderation code. Please add it back in. 

On 18 December 2023 00:05:53 GMT, Ilya Pekert ***@***.***> wrote:
>@misterpekert commented on this pull request.
>
>
>
>>    },
>-  confirm: {
>
>Well, for me it feels that having the confirmation modal as part of a spinning button a bit "overkill". Any part of the application might need confirmation flow, so from my point of view it should be something standalone. We can think of `<SpinButtonWithConfirm>` component which would have all the perks of SpinButton + confirmation popup, but it felt a bit too much considering the only place of use was `IsoChrone` component.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/Freegle/iznik-nuxt3/pull/47#discussion_r1429319057
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

misterpekert commented on components/SpinButton.vue:
```diff
@@ -1,124 +1,116 @@
-    setTimeout(() => {
-      done.value = false
-    }, props.timeout)
+    emit('handle', finnishSpinner)
```
Yeah, it would be spinning until callback is called. Adding 20 sec timer + Sentry should be enough to catch places where we forget to call the callback, on it.

misterpekert replied:
Fixed that, somehow slipped through :(

edwh replied:
Ok - please do that.

misterpekert commented on components/MessageEditModal.vue:
```diff
@@ -235,7 +234,6 @@ export default {
   methods: {
     async save() {
       this.triedToSave = true
-      if (this.isSaveButtonDisabled) return
```
No, it's safe to assume that if button is disabled, no click events are dispatched.

misterpekert replied:
Oh, yeah, my bad. I've added this for some testing and forgot to remove. SpinButton would never be rendered if autoSave is turned on, so this watch should be left intact. Rolling back that part of the change.

misterpekert commented on components/PostCode.vue:
```diff
@@ -49,17 +49,17 @@
         />
         <div v-if="find && !wip">
-          <b-button
+          <SpinButton
```
One way would be a transition, another would be to have a minimum of, say, 1 second of duration for that spinner. Even if the action is immediate, ain't seeing anything bad in showing spinner for at least a second. (same effect can be achieved with CSS transition) Extracted here -> #60

misterpekert replied:
Fixed that, somehow slipped through :(

misterpekert commented on components/SpinButton.vue:
```diff
@@ -1,124 +1,116 @@
-        <span v-if="name">
-          <v-icon v-if="done" icon="check" :class="spinclass + ' fa-fw'" />
-          <v-icon
-            v-else-if="doing"
```
One way would be a transition, another would be to have a minimum of, say, 1 second of duration for that spinner. Even if the action is immediate, ain't seeing anything bad in showing spinner for at least a second. (same effect can be achieved with CSS transition) Extracted here -> #60

edwh replied:
Thanks for opening the issue.

misterpekert commented:
> I've found it very useful to have that function to easily add or remove the confirmation stage, and as I say, it's used in the moderation code. Please add it back in.
> […](#)
> On 18 December 2023 00:05:53 GMT, Ilya Pekert ***@***.***> wrote: @misterpekert commented on this pull request. > }, - confirm: { Well, for me it feels that having the confirmation modal as part of a spinning button a bit "overkill". Any part of the application might need confirmation flow, so from my point of view it should be something standalone. We can think of `<SpinButtonWithConfirm>` component which would have all the perks of SpinButton + confirmation popup, but it felt a bit too much considering the only place of use was `IsoChrone` component. -- Reply to this email directly or view it on GitHub: [#47 (comment)](https://github.com/Freegle/iznik-nuxt3/pull/47#discussion_r1429319057) You are receiving this because you were mentioned. Message ID: ***@***.***>

`confirm` prop and modal brought back.

edwh commented:
The method won't throw an exception. But access to navigator.geolocation is the kind of thing that gets blocked by privacy blockers, and I've seen all sorts of weird exceptions in weird browsers with weird privacy blockers over the years. 

I want to avoid that being an issue, so please handle the exception case by setting locating back to false even if you don't think it's needed. Thanks. 

On 21 December 2023 10:11:27 GMT, Ilya Pekert ***@***.***> wrote:
>@misterpekert commented on this pull request.
>
>
>
>> @@ -116,6 +118,7 @@ export default {
>           this.locating = true
>           navigator.geolocation.getCurrentPosition((position) => {
>             // Show close to where we think they are.
>+            this.locating = false;
>
>I'm notsure what you mean, `navigator.geolocation.getCurrentPosition` has two arguments: `successCallback` and `errorCallback`. I've added `this.locating = false` to both of them. But now, when we don't need this flag, it's enough to call the `callback` method. I've updated the code. I'm not sure what the logic was before, but `navigator.geolocation.getCurrentPosition` never throws an error, so I've added the `this.locationFailed = true` part to the `errorCallback`. Let me know if we should remove try/catch from here.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/Freegle/iznik-nuxt3/pull/47#discussion_r1433858012
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
